### PR TITLE
Implement third party identifier invites

### PR DESCRIPTION
**Pull request by illicitonion**
As spec'd in https://github.com/matrix-org/matrix-doc/pull/41
And tested in https://github.com/matrix-org/matrix-doc/pull/41

### CODE REVIEW

File: synapse/api/auth.py [hunk 1]
```diff
@@ -367,6 +373,31 @@ def is_membership_change_allowed(self, event, auth_events):
 
         return True
 
+    def _verify_third_party_invite(self, event, auth_events):
```
Comment (erikjohnston) [inline synapse/api/auth.py]: Could we have a quick docstring/comment here to outline what this actually checks?

File: synapse/api/auth.py [hunk 2]
```diff
@@ -646,6 +677,14 @@ def compute_auth_events(self, event, current_state):
...
                 if member_event and not is_public:
                     auth_ids.append(member_event.event_id)
+                if ThirdPartyInvites.join_has_third_party_invite(event.content):
+                    key = (
+                        EventTypes.ThirdPartyInvite,
+                        event.content["third_party_invite"]["token"]
+                    )
+                    invite = current_state.get(key)
+                    if invite:
+                        auth_ids.append(invite.event_id)
```
Comment (erikjohnston) [inline synapse/api/auth.py]: Do we want to explode if we *don't* have that invite?

File: synapse/handlers/federation.py [hunk 3]
```diff
@@ -162,7 +162,7 @@ def on_receive_pdu(self, origin, pdu, backfilled, state=None,
                     auth_ids = [e_id for e_id, _ in e.auth_events]
                     auth = {
                         (e.type, e.state_key): e for e in auth_chain
-                        if e.event_id in auth_ids or e.type == EventTypes.Create
+                        if e.event_id in auth_ids
                     }
```
Comment (erikjohnston) [inline synapse/handlers/federation.py]: Why is this removed?

File: synapse/rest/client/v1/room.py [hunk 4]
```diff
@@ -440,6 +474,94 @@ def on_POST(self, request, room_id, membership_action, txn_id=None):
...
+            data = yield self.hs.get_simple_http_client().get_json(
+                "https://%s/_matrix/identity/api/v1/lookup" % (id_server,),
+                {
+                    "medium": medium,
+                    "address": address,
+                }
+            )
+
+            if "mxid" in data:
+                # TODO: Validate the response signature and such
```
Comment (erikjohnston) [inline synapse/rest/client/v1/room.py]: Do we verify ID server signatures elsewhere?

File: synapse/rest/client/v1/room.py [hunk 5]
```diff
@@ -440,6 +474,94 @@ def on_POST(self, request, room_id, membership_action, txn_id=None):
...
+                defer.returnValue(data["mxid"])
+        except IOError:
+            # TODO: Log something maybe?
```
Comment (erikjohnston) [inline synapse/rest/client/v1/room.py]: Yes :) Maybe a log.warn?

File: synapse/rest/client/v1/room.py [hunk 6]
```diff
@@ -415,9 +415,34 @@ def on_POST(self, request, room_id, membership_action, txn_id=None):
...
+                        display_name,
+                        medium,
+                        address,
+                        room_id,
+                        user,
+                        token_id,
+                        txn_id=txn_id
+                    )
+                    defer.returnValue((200, {}))
+                    return
```
Comment (erikjohnston) [inline synapse/rest/client/v1/room.py]: This feels like logic that shouldn't be in the rest layer.

File: synapse/rest/client/v1/room.py [hunk 7]
```diff
@@ -440,6 +474,94 @@ def on_POST(self, request, room_id, membership_action, txn_id=None):
...
+                "state_key": token,
+            },
+            token_id=token_id,
+            txn_id=txn_id,
+        )
+
+    @defer.inlineCallbacks
+    def _ask_id_server_for_third_party_invite(
+            self, id_server, medium, address, room_id, sender):
+        is_url = "https://%s/_matrix/identity/api/v1/nonce-it-up" % (id_server,)
```
Comment (erikjohnston) [inline synapse/rest/client/v1/room.py]: Is `nonce-it-up` really what we're calling it?

File: synapse/util/thirdpartyinvites.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from twisted.internet import defer
+from synapse.api.errors import AuthError
+
+
+class ThirdPartyInvites(object):
```
Comment (erikjohnston) [inline synapse/util/thirdpartyinvites.py]: Is there a reason why we have this class? Python supports free functions quite happily.

File: synapse/api/auth.py (hunk 1 above)
Comment (illicitonion) [inline synapse/api/auth.py]: Done

File: synapse/api/auth.py (hunk 2 above)
Comment (illicitonion) [inline synapse/api/auth.py]: Nope; this is just a logic method to tell you whether there is a thing being satisfied; the caller is responsible for handling that how it will.

File: synapse/handlers/federation.py (hunk 3 above)
Comment (illicitonion) [inline synapse/handlers/federation.py]: Good question. Re-added. Shame it didn't make any tests fail...

File: synapse/util/thirdpartyinvites.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    def join_has_third_party_invite(cls, content):
+        if "third_party_invite" not in content:
+            return False
+        return cls.has_join_keys(content["third_party_invite"])
+
+    @classmethod
+    def copy_join_keys(cls, src, dst):
+        for key in cls.JOIN_KEYS:
+            if key in src:
+                dst[key] = src[key]
```
Comment (erikjohnston) [inline synapse/util/thirdpartyinvites.py]: Personally, I would write:

```python
def extract_join_keys(src):
    return {key: value for key, value in src.items() if key in JOIN_KEYS}
```

And then the call site can do either of:

```python
existing_dict.update(extract_join_keys(src))   # 1.
new_dict = extract_join_keys(src)              # 2.

File: synapse/util/thirdpartyinvites.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    def check_key_valid(cls, http_client, event):
+        try:
+            response = yield http_client.get_json(
+                event.content["third_party_invite"]["key_validity_url"],
+                {"public_key": event.content["third_party_invite"]["public_key"]}
+            )
+            if not response["valid"]:
+                raise AuthError(403, "Third party certificate was invalid")
+        except IOError:
+            raise AuthError(403, "Third party certificate could not be checked")
```
Comment (erikjohnston) [inline synapse/util/thirdpartyinvites.py]: Is `403` the correct error here? It may be worth retrying the request later.

File: synapse/api/auth.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/api/auth.py]: Isn't it used to fill out the `auth_events` field in events?

File: synapse/util/thirdpartyinvites.py (hunk 8 above)
Comment (illicitonion) [inline synapse/util/thirdpartyinvites.py]: Done

File: synapse/util/thirdpartyinvites.py (hunk 9 above)
Comment (illicitonion) [inline synapse/util/thirdpartyinvites.py]: Done

File: synapse/util/thirdpartyinvites.py (hunk 10 above)
Comment (illicitonion) [inline synapse/util/thirdpartyinvites.py]: Done

File: synapse/rest/client/v1/room.py (hunk 6 above)
Comment (illicitonion) [inline synapse/rest/client/v1/room.py]: Done

File: synapse/rest/client/v1/room.py (hunk 5 above)
Comment (illicitonion) [inline synapse/rest/client/v1/room.py]: Done

**Comment by erikjohnston** [comment]
Also, we probably should be specifying what type of signature we're using, so that we can change it as necessary down the line.

File: synapse/rest/client/v1/room.py (hunk 7 above)
Comment (illicitonion) [inline synapse/rest/client/v1/room.py]: Aah yes, I forgot all about this placeholder. Done.

File: synapse/rest/client/v1/room.py (hunk 4 above)
Comment (illicitonion) [inline synapse/rest/client/v1/room.py]: Of course not. We do here, now, though :)

File: synapse/api/auth.py (hunk 2 above)
Comment (illicitonion) [inline synapse/api/auth.py]: We'll explode when we try to look up the value in the check, later.

Comment (erikjohnston) [inline synapse/api/auth.py]: Aha, yes. Fair enough.

**Comment by illicitonion** [comment]
Added key name structure as we do everywhere else.

Will update spec when you're happy with it.

**Comment by erikjohnston** [comment]
LGTM
